[new release] alt-ergo (4 packages) (2.5.0)

Halbaroth opened the pull request:
The Alt-Ergo SMT prover

- Project page: <a href="https://alt-ergo.ocamlpro.com/">https://alt-ergo.ocamlpro.com/</a>
- Documentation: <a href="https://ocamlpro.github.io/alt-ergo">https://ocamlpro.github.io/alt-ergo</a>

kit-ty-kate commented on packages/alt-ergo-lib/alt-ergo-lib.2.5.0/opam:
```diff
@@ -0,0 +1,63 @@
+  "seq"
+  "fmt"
+  "stdlib-shims"
+  "ppx_blob" {>= "0.3"}
```
I'm unsure which version to require but I'm guessing the first version to have moved to ppxlib is probably a good bet:
```
#=== ERROR while compiling alt-ergo-lib.2.5.0 =================================#
# context              2.2.0~alpha3~dev | linux/x86_64 | ocaml-base-compiler.4.08.1 | file:///home/opam/opam-repository
# path                 ~/.opam/4.08/.opam-switch/build/alt-ergo-lib.2.5.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p alt-ergo-lib -j 31 --promote-install-files=false @install
# exit-code            1
# env-file             ~/.opam/log/alt-ergo-lib-7-5588cb.env
# output-file          ~/.opam/log/alt-ergo-lib-7-5588cb.out
### output ###
# File "/home/opam/.opam/4.08/lib/ppx_blob/ppx_blob.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# File "/home/opam/.opam/4.08/lib/ppx_derivers/ppx_derivers.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# File "src/lib/dune", line 28, characters 19-27:
# 28 |   (preprocess (pps ppx_blob))
#                         ^^^^^^^^
# Error: Ppx dependency on a non-ppx library "ppx_blob". If "ppx_blob" is in
# fact a ppx rewriter library, it should have (kind ppx_rewriter) in its dune
# file.
# -> required by library "alt-ergo-lib" in _build/default/src/lib
# -> required by _build/default/META.alt-ergo-lib
# -> required by _build/install/default/lib/alt-ergo-lib/META
# -> required by _build/default/alt-ergo-lib.install
# -> required by alias install
# File "src/lib/dune", line 28, characters 14-28:
# 28 |   (preprocess (pps ppx_blob))
#                    ^^^^^^^^^^^^^^
# (cd _build/default && .ppx/c3461e88ae1503d575d582eb5874e916/ppx.exe --cookie 'library-name="AltErgoLib"' -o src/lib/reasoners/bitv.pp.ml --impl src/lib/reasoners/bitv.ml --dump-ast)
# Fatal error: exception File "src/lib/reasoners/bitv.ml", line 225, characters 6-10: let operators are not supported before OCaml 4.08
# (cd _build/default && .ppx/c3461e88ae1503d575d582eb5874e916/ppx.exe --cookie 'library-name="AltErgoLib"' -o src/lib/util/theories.pp.ml --impl src/lib/util/theories.ml --dump-ast)
# File "src/lib/util/theories.ml", line 38, characters 18-39:
# 38 |   | Fpa -> [%blob "src/preludes/fpa.ae"]
#                        ^^^^^^^^^^^^^^^^^^^^^
# Error: [%blob] could not find or load file /workspace_root/src/lib/util/src/preludes/fpa.ae
```
```suggestion
  "ppx_blob" {>= "0.7.2"}
```

kit-ty-kate commented:
The tests on alt-ergo-lib seem to run the tests meant for alt-ergo:
```
#=== ERROR while compiling alt-ergo-lib.2.5.0 =================================#
# context              2.2.0~alpha3~dev | linux/x86_64 | ocaml-base-compiler.4.11.2 | pinned(https://github.com/OCamlPro/alt-ergo/releases/download/v2.5.0/alt-ergo-2.5.0.tbz)
# path                 ~/.opam/4.11/.opam-switch/build/alt-ergo-lib.2.5.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p alt-ergo-lib -j 255 --promote-install-files=false @install @runtest
# exit-code            1
# env-file             ~/.opam/log/alt-ergo-lib-7-d8d080.env
# output-file          ~/.opam/log/alt-ergo-lib-7-d8d080.out
### output ###
# File "src/plugins/AB-Why3/preludes/cram.t/run.t", line 1, characters 0-0:
# /usr/bin/git --no-pager diff --no-index --color=always -u _build/.sandbox/958109ad212ad5e858b9df4af5c6ccdf/default/src/plugins/AB-Why3/preludes/cram.t/run.t _build/.sandbox/958109ad212ad5e858b9df4af5c6ccdf/default/src/plugins/AB-Why3/preludes/cram.t/run.t.corrected
# diff --git a/_build/.sandbox/958109ad212ad5e858b9df4af5c6ccdf/default/src/plugins/AB-Why3/preludes/cram.t/run.t b/_build/.sandbox/958109ad212ad5e858b9df4af5c6ccdf/default/src/plugins/AB-Why3/preludes/cram.t/run.t.corrected
# index 30966bd..4fe6ca9 100644
# --- a/_build/.sandbox/958109ad212ad5e858b9df4af5c6ccdf/default/src/plugins/AB-Why3/preludes/cram.t/run.t
# +++ b/_build/.sandbox/958109ad212ad5e858b9df4af5c6ccdf/default/src/plugins/AB-Why3/preludes/cram.t/run.t.corrected
# @@ -1,3 +1,7 @@
#    $ alt-ergo ../b-set-theory-prelude-2018-09-28.ae
# +  alt-ergo: not found
# +  [127]
#    $ alt-ergo ../b-set-theory-prelude-2020-02-28.ae
# +  alt-ergo: not found
# +  [127]
#  
# File "src/preludes/cram.t/run.t", line 1, characters 0-0:
# /usr/bin/git --no-pager diff --no-index --color=always -u _build/.sandbox/e19a84018fb545aa547e52345013d34c/default/src/preludes/cram.t/run.t _build/.sandbox/e19a84018fb545aa547e52345013d34c/default/src/preludes/cram.t/run.t.corrected
# diff --git a/_build/.sandbox/e19a84018fb545aa547e52345013d34c/default/src/preludes/cram.t/run.t b/_build/.sandbox/e19a84018fb545aa547e52345013d34c/default/src/preludes/cram.t/run.t.corrected
# index f7339f0..33f6ba3 100644
# --- a/_build/.sandbox/e19a84018fb545aa547e52345013d34c/default/src/preludes/cram.t/run.t
# +++ b/_build/.sandbox/e19a84018fb545aa547e52345013d34c/default/src/preludes/cram.t/run.t.corrected
# @@ -1,3 +1,9 @@
#    $ alt-ergo ../fpa-theory-2017-01-04-16h00.ae
# +  alt-ergo: not found
# +  [127]
#    $ alt-ergo ../fpa-theory-2019-06-14-11h00.ae
# +  alt-ergo: not found
# +  [127]
#    $ alt-ergo ../fpa-theory-2019-10-08-19h00.ae
# +  alt-ergo: not found
# +  [127]
```
I'm guessing there is a missing `(package alt-ergo)` somewhere in your tests?

mseri commented on packages/alt-ergo/alt-ergo.2.5.0/opam:
```diff
@@ -0,0 +1,55 @@
+  "dune-site"
+  "cmdliner" {>= "1.1.0"}
+  "odoc" {with-doc}
+]
```
```suggestion
]
conflicts: [
  "result" {< "1.5"}
]
```

should resolve
```
#=== ERROR while compiling alt-ergo.2.5.0 =====================================#
# context              2.2.0~alpha3~dev | linux/x86_64 | ocaml-base-compiler.4.08.1 | file:///home/opam/opam-repository
# path                 ~/.opam/4.08/.opam-switch/build/alt-ergo.2.5.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p alt-ergo -j 31 --promote-install-files=false @install
# exit-code            1
# env-file             ~/.opam/log/alt-ergo-7-737784.env
# output-file          ~/.opam/log/alt-ergo-7-737784.out
### output ###
# (cd _build/default && /home/opam/.opam/4.08/bin/ocamlc.opt -w -40 -bin-annot -w -22 -g -bin-annot -I src/bin/common/.alt_ergo_common.objs/byte -I /home/opam/.opam/4.08/lib/alt-ergo-lib -I /home/opam/.opam/4.08/lib/alt-ergo-lib/__private__/alt_ergo_prelude/.public_cmi -I /home/opam/.opam/4.08/lib/alt-ergo-parsers -I /home/opam/.opam/4.08/lib/camlzip/../zip -I /home/opam/.opam/4.08/lib/cmdliner -I /home/opam/.opam/4.08/lib/dolmen -I /home/opam/.opam/4.08/lib/dolmen/ae -I /home/opam/.opam/4.08/lib/dolmen/class -I /home/opam/.opam/4.08/lib/dolmen/dimacs -I /home/opam/.opam/4.08/lib/dolmen/icnf -I /home/opam/.opam/4.08/lib/dolmen/intf -I /home/opam/.opam/4.08/lib/dolmen/line -I /home/opam/.opam/4.08/lib/dolmen/smtlib2 -I /home/opam/.opam/4.08/lib/dolmen/smtlib2/poly -I /home/opam/.opam/4.08/lib/dolmen/smtlib2/v6 -I /home/opam/.opam/4.08/lib/dolmen/smtlib2/v6_response -I /home/opam/.opam/4.08/lib/dolmen/smtlib2/v6_script -I /home/opam/.opam/4.08/lib/dolmen/std -I /home/opam/.opam/4.08/lib/dolmen/tptp -I /home/opam/.opam/4.08/lib/dolmen/tptp/v6_3_0 -I /home/opam/.opam/4.08/lib/dolmen/zf -I /home/opam/.opam/4.08/lib/dolmen_loop -I /home/opam/.opam/4.08/lib/dolmen_type -I /home/opam/.opam/4.08/lib/dune-build-info -I /home/opam/.opam/4.08/lib/dune-private-libs/dune-section -I /home/opam/.opam/4.08/lib/dune-site -I /home/opam/.opam/4.08/lib/fmt -I /home/opam/.opam/4.08/lib/gen -I /home/opam/.opam/4.08/lib/logs -I /home/opam/.opam/4.08/lib/menhirLib -I /home/opam/.opam/4.08/lib/ocplib-simplex -I /home/opam/.opam/4.08/lib/pp_loc -I /home/opam/.opam/4.08/lib/psmt2-frontend -I /home/opam/.opam/4.08/lib/result -I /home/opam/.opam/4.08/lib/seq -I /home/opam/.opam/4.08/lib/spelll -I /home/opam/.opam/4.08/lib/stdlib-shims -I /home/opam/.opam/4.08/lib/uchar -I /home/opam/.opam/4.08/lib/uutf -I /home/opam/.opam/4.08/lib/zarith -intf-suffix .ml -no-alias-deps -open Alt_ergo_common -o src/bin/common/.alt_ergo_common.objs/byte/alt_ergo_common__Parse_command.cmo -c -impl src/bin/common/parse_command.ml)
# File "src/bin/common/parse_command.ml", line 645, characters 11-22:
# 645 |            Result.bind res (fun () -> load_parser verbose path))
#                  ^^^^^^^^^^^
# Error: Unbound value Result.bind
# (cd _build/default && /home/opam/.opam/4.08/bin/ocamlopt.opt -w -40 -bin-annot -w -22 -O3 -unbox-closures -I src/bin/common/.alt_ergo_common.objs/byte -I src/bin/common/.alt_ergo_common.objs/native -I /home/opam/.opam/4.08/lib/alt-ergo-lib -I /home/opam/.opam/4.08/lib/alt-ergo-lib/__private__/alt_ergo_prelude -I /home/opam/.opam/4.08/lib/alt-ergo-lib/__private__/alt_ergo_prelude/.public_cmi -I /home/opam/.opam/4.08/lib/alt-ergo-parsers -I /home/opam/.opam/4.08/lib/camlzip/../zip -I /home/opam/.opam/4.08/lib/cmdliner -I /home/opam/.opam/4.08/lib/dolmen -I /home/opam/.opam/4.08/lib/dolmen/ae -I /home/opam/.opam/4.08/lib/dolmen/class -I /home/opam/.opam/4.08/lib/dolmen/dimacs -I /home/opam/.opam/4.08/lib/dolmen/icnf -I /home/opam/.opam/4.08/lib/dolmen/intf -I /home/opam/.opam/4.08/lib/dolmen/line -I /home/opam/.opam/4.08/lib/dolmen/smtlib2 -I /home/opam/.opam/4.08/lib/dolmen/smtlib2/poly -I /home/opam/.opam/4.08/lib/dolmen/smtlib2/v6 -I /home/opam/.opam/4.08/lib/dolmen/smtlib2/v6_response -I /home/opam/.opam/4.08/lib/dolmen/smtlib2/v6_script -I /home/opam/.opam/4.08/lib/dolmen/std -I /home/opam/.opam/4.08/lib/dolmen/tptp -I /home/opam/.opam/4.08/lib/dolmen/tptp/v6_3_0 -I /home/opam/.opam/4.08/lib/dolmen/zf -I /home/opam/.opam/4.08/lib/dolmen_loop -I /home/opam/.opam/4.08/lib/dolmen_type -I /home/opam/.opam/4.08/lib/dune-build-info -I /home/opam/.opam/4.08/lib/dune-private-libs/dune-section -I /home/opam/.opam/4.08/lib/dune-site -I /home/opam/.opam/4.08/lib/fmt -I /home/opam/.opam/4.08/lib/gen -I /home/opam/.opam/4.08/lib/logs -I /home/opam/.opam/4.08/lib/menhirLib -I /home/opam/.opam/4.08/lib/ocplib-simplex -I /home/opam/.opam/4.08/lib/pp_loc -I /home/opam/.opam/4.08/lib/psmt2-frontend -I /home/opam/.opam/4.08/lib/result -I /home/opam/.opam/4.08/lib/seq -I /home/opam/.opam/4.08/lib/spelll -I /home/opam/.opam/4.08/lib/stdlib-shims -I /home/opam/.opam/4.08/lib/uchar -I /home/opam/.opam/4.08/lib/uutf -I /home/opam/.opam/4.08/lib/zarith -intf-suffix .ml -no-alias-deps -open Alt_ergo_common -o src/bin/common/.alt_ergo_common.objs/native/alt_ergo_common__Parse_command.cmx -c -impl src/bin/common/parse_command.ml)
# File "src/bin/common/parse_command.ml", line 645, characters 11-22:
# 645 |            Result.bind res (fun () -> load_parser verbose path))
#                  ^^^^^^^^^^^
# Error: Unbound value Result.bind
```

mseri commented:
Do you understand why sometimes many versions of framac (24 -> 27) fail to compile in some cases? E.g. https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/90f211d0c95a3b8d44b83cbb8391351a5cc34c07/variant/compilers,4.14,alt-ergo.2.5.0,revdeps,frama-c.24.0

Halbaroth commented:
It seems a printer have changed in the codebase of frama-c but new test outputs haven't been promoted yet. I will contact the frama-c team.

AllanBlanchard commented:
The problem is due to an incompatibility with the version 2.1.0 of OCamlGraph merged two weeks ago: https://github.com/ocaml/opam-repository/pull/24333 . The CI had technical difficulties so the problem could not be spotted. For tests, all versions of Frama-C fail with OCamlGraph 2.1.0 (commit : https://github.com/backtracking/ocamlgraph/commit/48c2ae062edbdbe6106a0a5ce8df353372e953ba) because the new end-of-line that appears when displaying graphs in dot.

mseri commented:
Thanks a lot for looking into it. As soon as I (or somebody seeing this) have a moment, I'll add a bound for the tests of frama-c 24 - 27.1 and then re-run the failed CI tests here.

Halbaroth commented:
Thanks for your help Allan :)

mseri commented:
Thanks a lot!

mseri commented:
Do you plan to address the failures on archlinux?
```
#=== ERROR while compiling alt-ergo.2.5.0 =====================================#
# context              2.2.0~alpha3~dev | linux/x86_64 | ocaml-base-compiler.4.14.1 | pinned(https://github.com/OCamlPro/alt-ergo/releases/download/v2.5.0/alt-ergo-2.5.0.tbz)
# path                 ~/.opam/4.14/.opam-switch/build/alt-ergo.2.5.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p alt-ergo -j 127 --promote-install-files=false @install @runtest
# exit-code            1
# env-file             ~/.opam/log/alt-ergo-7-0eb299.env
# output-file          ~/.opam/log/alt-ergo-7-0eb299.out
### output ###
# File "tests/cram.t/run.t", line 1, characters 0-0:
# /usr/sbin/git --no-pager diff --no-index --color=always -u _build/.sandbox/b8043f736682073678531f6027826151/default/tests/cram.t/run.t _build/.sandbox/b8043f736682073678531f6027826151/default/tests/cram.t/run.t.corrected
# diff --git a/_build/.sandbox/b8043f736682073678531f6027826151/default/tests/cram.t/run.t b/_build/.sandbox/b8043f736682073678531f6027826151/default/tests/cram.t/run.t.corrected
# index b121119..fcc395b 100644
# --- a/_build/.sandbox/b8043f736682073678531f6027826151/default/tests/cram.t/run.t
# +++ b/_build/.sandbox/b8043f736682073678531f6027826151/default/tests/cram.t/run.t.corrected
# @@ -1,4 +1,13 @@
#    $ export BUILD_PATH_PREFIX_MAP="SRC=$(dirname $(dirname $(dirname $(dirname $(which alt-ergo))))):$BUILD_PATH_PREFIX_MAP"
# +  SRC/opam-tmp/dune_cram_39018c_.cram.t/1.sh: line 1: which: command not found
# +  dirname: missing operand
# +  Try 'dirname --help' for more information.
# +  dirname: missing operand
# +  Try 'dirname --help' for more information.
# +  dirname: missing operand
# +  Try 'dirname --help' for more information.
# +  dirname: missing operand
# +  Try 'dirname --help' for more information.
#    $ echo '(check-sat)' | alt-ergo --inequalities-plugin does-not-exist -i smtlib2 -o smtlib2 2>&1 >/dev/null | sed -E 's/\(\\".*\\"\)//'
#    alt-ergo: ; Fatal Error: [Dynlink] Loading the 'inequalities' reasoner (FM module) plugin in "does-not-exist" failed!
#              >> Failure message: error loading shared library: Dynlink.Error (Dynlink.Cannot_open_dll "Failure")
```

Halbaroth commented:
> Do you plan to address the failures on archlinux?

I did. The cram test wasn't useful anymore. I just removed it. 
I also fixed other issues with ppxlib and Zarith. I hope the CI will pass now :)

mseri commented:
Thanks for the help and the patience!
